datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java

```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
       return _zkClient;
     }
 
+    public class ZkClientMockStateChangeListener extends ZkStateChangeListener {
+      boolean sessionExpired = false;
+      @Override
+      public void handleStateChanged(Watcher.Event.KeeperState state) {
+        super.handleStateChanged(state);
+        if (state == Watcher.Event.KeeperState.Expired) {
+          LOG.info("ZkStateChangeListener::Session expired.");
+          sessionExpired = true;
+        }
+      }
+    }
+
+    @Override
+    ZkStateChangeListener getOrCreateStateChangeListener() {
+      _zkClientMockStateChangeListener = spy(new ZkClientMockStateChangeListener());
+      return _zkClientMockStateChangeListener;
+    }
+
     public ZkClient getZkClient() {
       return _zkClient;
     }
+
+    public ZkClientMockStateChangeListener getZkStateChangeListener() {
+      return _zkClientMockStateChangeListener;
+    }
+  }
+
+  @Test
+  public void testZookeeperSessionExpiry() throws InterruptedException {
+    String testCluster = "testDeleteTaskWithPrefix";
+    String connectorType = "connectorType";
+    Duration timeout = Duration.ofMinutes(1);
+
+    ZkClientInterceptingAdapter adapter = createInterceptingZkAdapter(testCluster, 5000);
+    adapter.connect();
+
+    DatastreamTaskImpl task = new DatastreamTaskImpl();
+    task.setId("3");
+    task.setConnectorType(connectorType);
+    task.setZkAdapter(adapter);
+
+    List<DatastreamTask> tasks = new ArrayList<>();
+    tasks.add(task);
+    updateInstanceAssignment(adapter, adapter.getInstanceName(), tasks);
+
+    LOG.info("Acquire from instance1 should succeed");
+    Assert.assertTrue(expectException(() -> task.acquire(timeout), false));
+
+    simulateSessionExpiration(adapter);
+
+    Thread.sleep(5000);
```

Comment:
Please excuse my ignorance: why do we have to sleep for that long? Is it because this is the value we specified for `sessionTimeoutMs`? If that's the reason, should we decrease that value?

Reply:
I gave a decent value so that it does not fail on pcl machines which are slower and then this test will become flaky.